src/streams.rs

```diff
@@ -0,0 +1,669 @@
+                    .collect::<Vec<StreamKey>>()
+            })
+            .collect();
+        Ok(StreamReadReply { keys })
+    }
+}
+
+impl FromRedisValue for StreamRangeReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let ids: Vec<StreamId> = rows
+            .into_iter()
+            .flat_map(|row| {
+                row.into_iter()
+                    .map(|(id, map)| StreamId { id, map })
+                    .collect::<Vec<StreamId>>()
+            })
+            .collect();
+        Ok(StreamRangeReply { ids })
+    }
+}
+
+impl FromRedisValue for StreamClaimReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let ids: Vec<StreamId> = rows
+            .into_iter()
+            .flat_map(|row| {
+                row.into_iter()
+                    .map(|(id, map)| StreamId { id, map })
+                    .collect::<Vec<StreamId>>()
+            })
+            .collect();
+        Ok(StreamClaimReply { ids })
+    }
+}
+
+impl FromRedisValue for StreamPendingReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let (count, start, end, consumer_data): (
+            _,
+            Option<String>,
+            Option<String>,
+            Vec<Vec<String>>,
+        ) = from_redis_value(v)?;
+
+        if count == 0 {
+            Ok(StreamPendingReply::Empty)
+        } else {
+            let mut result = StreamPendingData::default();
+
+            let start_id = start.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects start id",
+                )
+            })?;
+
+            let end_id = end.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects end id",
+                )
+            })?;
+
+            result.count = count;
+            result.start_id = start_id;
+            result.end_id = end_id;
+
+            for consumer in consumer_data {
+                let mut info = StreamInfoConsumer::default();
+                info.name = consumer[0].to_owned();
+                if let Ok(v) = consumer[1].parse::<usize>() {
+                    info.pending = v;
+                }
+                result.consumers.push(info);
+            }
+
+            Ok(StreamPendingReply::Data(result))
+        }
+    }
+}
+
+impl FromRedisValue for StreamPendingCountReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let parts: Vec<Vec<(String, String, usize, usize)>> = from_redis_value(v)?;
+        let mut reply = StreamPendingCountReply::default();
+        for row in parts {
+            reply.ids.push(StreamPendingId {
+                id: row[0].0.to_owned(),
```

Comment:
How come this only extracts the first element of `row`? Is it always a 1-element sequence or are the other elements just ignored?

Reply:
Let me dig in on this, as I'm also worried about.

Reply:
It's always a one-element sequence.  I dug into the docs and the data ends up matching what's expected to be returned from redis, it's just going through some contortions while being processed in `FromRedisValue` -- the tuple being nested in an array causes a strange shape.  

Here's a sample print-out that I captured by adding in some `println` (temporarily) to the test suite:

```text
  [("1000-1", "c99", 3, 2)]
  [("1001-0", "c99", 1, 1)]
  [("1001-1", "c99", 1, 1)]
  [("1000-1", "c99", 4, 2)]
  [("1001-0", "c99", 2, 1)]
  [("1001-1", "c99", 2, 1)]
 ```

Too, I expanded the test coverage to make sure that EVERY row processed by that function has _all four_ expected values.  I'll update the branch with that additional test, now.

This is the [relevant page from the Redis documentation](https://redis.io/commands/xpending)....

...which is actually a pretty dense bit of information.  Here's the specific part that we're interested in:

![xpending_detail](https://user-images.githubusercontent.com/38859656/83685882-57459e80-a5b7-11ea-8f4c-31775ad66009.png)

Reply:
Ok, in that case you should be able to use a one element tuple instead of a `Vec`, `((A, B, C, D),)` instead of `Vec<(A, B, C, D)>`.

Reply:
I took your advice, but for some reason, the data structure ends up mangled.

https://github.com/Terkwood/redis-rs/blob/dc648209a72a68f85b52ee4f3cc11b9d6fba0afb/src/streams.rs#L590

Here's a temp branch.  If you run the tests against this branch, you can see that `parts` is reduced to length 2, incorrectly.  It should be length 3.

I tested using the following:
```sh
REDISRS_SERVER_TYPE=unix cargo test --all-features -- --skip test_cluster --nocapture
```

I've been trying several different minor changes and can't fathom why it's not coalescing the `Value::Bulk`  into a `Vec<((String, String, usize, usize),)>` as expected.

Reply:
(Please ignore the ugly/wasteful `id: (row.0).0.to_owned()` in that branch -- I realize we can dispense with the call to `.to_owned`, once this is figured out -- but am focused on the fact that what looked like an innocent refactor is actually breaking tests.)

Reply:
Here's the relevant test spam:

The initial `Value::Bulk` looks fine, just like we expect:
```text
bulk(bulk(string-data('"1000-1"'), string-data('"c99"'), int(2), int(2)), bulk(
string-data('"1001-0"'), string-data('"c99"'), int(1), int(1)), bulk(string-data('"1001
-1"'), string-data('"c99"'), int(1), int(1)))
```

But then we declare a nicely-shaped `parts`, which is `from_redis_value`d incorrectly:

`let parts: Vec<((String, String, usize, usize),)> = from_redis_value(v)?;`

This ends up turning out to be only size 2, not 3.  :warning: :fire: 

```
 [(("1000-1", "c99", 2, 2),), (("1001-0", "c99", 1, 1),)]

...SNIP...
thread 'test_assorted_2' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `3`', tests/test_streams.rs:315:5
```

Reply:
I believe this macro is being called as part of the processing.  Not sure if it has any bearing on the situation, as I'm still very unaccomplished with rust macros.

```rust
macro_rules! from_redis_value_for_tuple 
```

Reply:
At least by the time that macro is invoked, it only sees 2 of the 3 elements.  Anyway, please don't take this excessive spam as any sort of implied rush!  I'll dig in a bit more before I keep posting notes :grin:   Definitely something interesting going on here

Reply:
OK. I think you will find this interesting.  I hacked the macro and got all the tests to pass.  :smiling_imp: 

Look toward the bottom:   `while offset < items.len() /* - 1 */ {  // WE HACKED IT! :grin:`

```rust

macro_rules! from_redis_value_for_tuple {
    () => ();
    ($($name:ident,)+) => (
        #[doc(hidden)]
        impl<$($name: FromRedisValue),*> FromRedisValue for ($($name,)*) {
            // we have local variables named T1 as dummies and those
            // variables are unused.
            #[allow(non_snake_case, unused_variables)]
            fn from_redis_value(v: &Value) -> RedisResult<($($name,)*)> {
                println!("I AM A TUPLE {:?}", v);
                match *v {
                    Value::Bulk(ref items) => {
                        // hacky way to count the tuple size
                        let mut n = 0;
                        $(let $name = (); n += 1;)*
                        if items.len() != n {
                            invalid_type_error!(v, "Bulk response of wrong dimension")
                        }

                        // this is pretty ugly too.  The { i += 1; i - 1} is rust's
                        // postfix increment :)
                        let mut i = 0;
                        Ok(($({let $name = (); from_redis_value(
                             &items[{ i += 1; i - 1 }])?},)*))
                    }
                    _ => invalid_type_error!(v, "Not a bulk response")
                }
            }

            #[allow(non_snake_case, unused_variables)]
            fn from_redis_values(items: &[Value]) -> RedisResult<Vec<($($name,)*)>> {
                // hacky way to count the tuple size
                let mut n = 0;
                $(let $name = (); n += 1;)*
                if items.len() % n != 0 {
                    invalid_type_error!(items, "Bulk response of wrong dimension")
                }

                // this is pretty ugly too.  The { i += 1; i - 1} is rust's
                // postfix increment :)
                let mut rv = vec![];
                if items.len() == 0 {
                    return Ok(rv)
                }
                let mut offset = 0;
                while offset < items.len() /* - 1 */ {  // WE HACKED IT! :grin:
                    rv.push(($({let $name = (); from_redis_value(
                         &items[{ offset += 1; offset - 1 }])?},)*));
                }
                Ok(rv)
            }
        }
        from_redis_value_for_tuple_peel!($($name,)*);
    )
}```

Reply:
Raised a PR for that change, specifically: #328

Reply:
I created a branch to show that #328 and #319 can play nicely together, by merging the changes from #328 into the changes here.

This is the relevant commit from that branch:  https://github.com/Terkwood/redis-rs/commit/7c0d5e6bfc1cc195844fc06f5a2da1f5ff097294

Reply:
OK, with #328 merged I'm still running into issues.  

```text
thread 'test_assorted_2' panicked at 'called `Result::unwrap()` on an `Err` value: Response was of incompatible type: "Bulk response of wrong dimension" (response was bulk(string-data('"1000-1"'), string-data('"c99"'), int(2), int(2)))', tests/test_streams.rs:314:42
```

This is due to incorrect counting in the relevant tuple macro:

```rust
// hacky way to count the tuple size
let mut n = 0;
 $(let $name = (); n += 1;)*
  if items.len() % n != 0 {
    invalid_type_error!(items, "Bulk response of wrong dimension")
 }
```

It can be proven by adding a test:

```rust

#[test]
fn test_vec_nested_tuple() {
    use redis::{FromRedisValue, Value};

    let v = FromRedisValue::from_redis_value(&Value::Bulk(vec![
        Value::Bulk(vec![
            Value::Data("1".into()),
            Value::Data("2".into()),
            Value::Data("hi".into()),
            Value::Data("redis".into()),
            Value::Data("99".into()),
        ]),
        Value::Bulk(vec![
            Value::Data("3".into()),
            Value::Data("4".into()),
            Value::Data("hello".into()),
            Value::Data("rust".into()),
            Value::Data("100".into()),
        ]),
    ]))
    .expect("from");

    let w = FromRedisValue::from_redis_value(&v);

    assert_eq!(
        w,
        Ok(vec![
            (1i32, 2, "hi".to_string(), "redis".to_string(), 99),
            (3i32, 4, "hello".to_string(), "rust".to_string(), 100)
        ])
    );
}
```

Within the macro itself, I can see that `n` is incorrectly counted as 2 (the length of the OUTER vec), when it should actually be counted as 5 (the size of the inner tuples).

Apparently I didn't run tests on the experimental branch (https://github.com/Terkwood/redis-rs/commit/7c0d5e6bfc1cc195844fc06f5a2da1f5ff097294), or would have caught this ... :facepalm:   Sorry about that!

The macro-writing exceeds my current level of understanding, so I'm going to need to brush up a bit more on how the macro system works before I can contribute a very useful edit.

Let me know how you'd like to proceed?

Reply:
It's worth noting that even with the improved version of tuple handling in #328 (shallow tuples at least get their length counted correctly, which previously wasn't the case), the test I posted in my previous comment shows that you need to call `from_redis_value` *twice*. 

This is due to the tuple being nested inside a variable-length `Vec`.

So I'm not sure that calling `from_redis_value` twice is any better than the `row[0]` version that currently used in *this* PR:

```rust
reply.ids.push(StreamPendingId {
                id: row[0].0.to_owned(),
                // ...SNIP
```

Reply:
Raised #334 so that the issue on nested tuples-inside-vecs doesn't get lost in the shuffle, regardless of what we choose to do here.

Personally, I'd vote for merging this PR with the various, sub-optimal `row[0].0.to_owned()` calls in tact, since we would gain access to high-level wrappers for Redis Streams.

We could then pursue #334 and circle back to this, to eventually remove the copies.  Getting the macro sorted out seems pretty complicated!

Reply:
Had another idea about this PR specifically. Maybe I can write an impl for FromRedisArgs which _only_ handles the `Vec<((String, String, usize, usize),),>` type, needed here. Will give it a try...

Reply:
OK!  I reworked this section completely.  Converted the PR back to draft status until I can fill out the remaining TODOs.  Should have those taken care of quickly.